src/inmanta/agent/handler.py

```diff
@@ -419,7 +419,52 @@ def _log_msg(self, level: int, msg: str, *args: object, exc_info: bool = False,
 
 
 @stable_api
-class ResourceHandler(object):
+class HandlerABC(ABC):
+    """
+    Top-level abstract base class all handlers should inherit from.
+    """
+
+    @abstractmethod
+    def pre(self, ctx: HandlerContext, resource: resources.Resource) -> None:
+        """
+        Method executed before a handler operation (Facts, dryrun, real deployment, ...) is executed. Override this method
+        to run before an operation.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to query facts for.
+        """
+
+    @abstractmethod
+    def post(self, ctx: HandlerContext, resource: resources.Resource) -> None:
+        """
+        Method executed after an operation. Override this method to run after an operation.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to query facts for.
+        """
+
+    @abstractmethod
+    def deploy(
+        self,
+        ctx: HandlerContext,
+        resource: resources.Resource,
+        requires: Dict[ResourceIdStr, ResourceState],
+    ) -> None:
+        """
+        This method is always called by the agent
+        """
+        pass
+
+    @abstractmethod
+    def execute(self, ctx: HandlerContext, resource: resources.Resource, dry_run: Optional[bool] = None) -> None:
+        """
+        Update the given resource. This method is called by the agent.
+        """
+        pass
```

Comment:
The docstrings on these methods have to be improved. What is the behavioral difference between `deploy()` and `execute()` for example? It's not possible to derive that information for the name itself.

Reply:
I agree, I was planning on hoisting these methods from `ResourceHandler` into `HandlerABC`, but now that I've worked on the actual implementation of the `DiscoveryHandler` I don't think I'll move these up after all.

Reply:
Perhaps `execute` shouldn't exist on the `ABC` then?

Reply:
Agreed, and perhaps neither should `deploy`, I'll update this accordingly after feedback on the PR implementing the `DiscoveryHandler`